src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java

```diff
@@ -61,20 +61,27 @@ DisplayURLProvider lookupProvider(StaplerRequest req) {
     }
 
     DisplayURLProvider lookupProvider() {
+        // We pick the first non-null option in this order:
+        // 1: PreferredProviderUserProperty (user preference)
+        // 2: DefaultProviderGlobalConfiguration (GUI-configured global default)
+        // 3: Env var/system properties defined in DisplayURLProvider (global default)
+        // 4: The DisplayUrlProvider extension with the highest ordinal value that is not an instance of ClassicDisplayURLProvider
+        // 5. ClassicDisplayURLProvider
         PreferredProviderUserProperty prefProperty = getUserPreferredProviderProperty();
 
         if (prefProperty != null && prefProperty.getConfiguredProvider() != null) {
             return prefProperty.getConfiguredProvider();
         }
-        DisplayURLProvider displayURLProvider = DisplayURLProvider.getPreferredProvider();
-        if (displayURLProvider == null) {
-            ExtensionList<DisplayURLProvider> all = DisplayURLProvider.all();
-            displayURLProvider = all.stream()
+        DisplayURLProvider globalProvider = DisplayURLProvider.getConfiguredDefault();
+        if (globalProvider != null) {
+            return globalProvider;
+        }
+        ExtensionList<DisplayURLProvider> all = DisplayURLProvider.all();
+        DisplayURLProvider displayURLProvider = all.stream()
                 .filter(
                     ((Predicate<DisplayURLProvider>) ClassicDisplayURLProvider.class::isInstance)
                         .negate())
                 .findFirst().orElse(DisplayURLProvider.getDefault());
```

Comment:
This old code is the bug behind [JENKINS-41578](https://issues.jenkins.io/browse/JENKINS-64119) / [JENKINS-41578](https://issues.jenkins.io/browse/JENKINS-64119). If you had _any_ non-`ClassicDisplayURLProvider` extension on your instance, it would take precedence over any environment variable/system property configured default when resolving `/display/redirect` URLs. #42 fixed that bug by never _generating_ `/display/redirect` URLs when the environment variable/system property is configured, instead hard-coding the URLs to the configured value, but that makes no sense as described in [JENKINS-69006](https://issues.jenkins.io/browse/JENKINS-69006).

The proper fix is to always generate `/display/redirect` URLs, but to consider the environment variable/system property configured default before using the ordinal-based fallback to pick a default provider.